torch/csrc/jit/passes/specialize_autogradzero.cpp

```diff
@@ -121,14 +213,19 @@ struct AutogradZeroSpecializer {
     std::vector<Value*> nonzero_values;
 
     for (auto inp : graph_->inputs()) {
-      if (auto profile_optional_node = getUse(inp, prim::profile_optional)) {
-        if (profile_optional_node->i(attr::num_present) == 0 &&
-            profile_optional_node->i(attr::num_none) != 0) {
+      if (auto profile_ivalue_node =
+              getUseWithAttribute(inp, countsAttribute)) {
+        TORCH_INTERNAL_ASSERT(
+            profile_ivalue_node->hasAttribute(countsAttribute));
+        const auto& counts_attr =
+            profile_ivalue_node->ival(countsAttribute).toGenericDict();
+        auto num_present = counts_attr.at(IValue{"num_present"}).toInt();
+        auto num_none = counts_attr.at(IValue{"num_none"}).toInt();
+        if (num_present == 0 && num_none != 0) {
           auto check = graph_->insert(aten::__is__, {inp, none_val})->node();
           checks.push_back(check->output());
           profiled_none_.insert(inp);
         }
-        removeProfiledOptionalUses(inp);
```

Comment:
I'm not convinced that changing `removeProfiledOptionalUses(inp)` to after `guardSpecializations` doesn't have some edge-case effect somewhere. Could we keep the code here as it was to minimize changes ?

Reply:
I think there were multiple reasons why I moved it after `guardSpecializations`. One of them was that there could be arbitrary long chains of `prim::profile`s on some graph inputs. Even for a very simple case I considered where both TE and SpecializeAutogradZero want to profile the optional list argument to `aten::grad_sum_to_size`, rewriting `removeProfiledOptionalUses` looked very ugly and I hit some edge cases there. 
In contrast, walking a graph and removing after `guardSpecializations` is done looked so much simpler and it also looked quite similar to what we do elsewhere.

I tested this with fastrnns and our CI tests and didn't hit any regressions or crashes.
If you feel very strongly that `removeProfiledOptionalUses(inp)` should stay in this specific spot, we could try figuring it out.

Reply:
If "rewriting removeProfiledOptionalUses looked very ugly" that is a good sign that pass writing is becoming more difficult as a result of the overloaded `profile_ivalue` cc @jjsjann123. 

i don't feel **that** strongly, but I do prefer it as it was. IMO when there isn't a large overhead of doing so, we should be cleaning the graph as we are manipulating it in the pass. It's the same reason I added https://github.com/pytorch/pytorch/blob/1b8d922d33f7e79d452cfcd109ad664e8cfd518a/torch/csrc/jit/passes/specialize_autogradzero.cpp#L301: it was extremely confusing to look at the pass midway through and see three copies of the graph. Why was it there ? Every time I debugged the pass I had to re-remind myself and mentally edit out the dead copy of the graph. Similarly, after we've made the decision whether or not to specialize on None values, why not remove the nodes right there? They are just visual noise for the duration of `guardSpecializations`. 

The logic right now isn't quite right, either. 
```
def foo(x):
   x = profile_ivalue[None=1, Present=1]
   return x
```
Here we won't insert any specializations, so the profile_ivalue nodes will remain in the graph. Granted this isn't likely to happen, but it's sort of a fallout of delaying logic past when it needs to be. 

I am maybe more opinionated about this than I should be because in the past I've had to deal with passes that had incorrect IR (like when we never updated the requires_grad property) or other passes that similarly left things in confusing states throughout the pass.